x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DelayedDataDetector.java

```diff
@@ -0,0 +1,144 @@
+import org.elasticsearch.xpack.core.ml.datafeed.DatafeedConfig;
+import org.elasticsearch.xpack.core.ml.datafeed.extractor.ExtractorUtils;
+import org.elasticsearch.xpack.core.ml.job.config.Job;
+import org.elasticsearch.xpack.core.ml.job.results.Bucket;
+import org.elasticsearch.xpack.core.ml.utils.Intervals;
+import org.joda.time.DateTime;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.stashWithOrigin;
+
+
+/**
+ * This class will search the buckets and indices over a given window to determine if any data is missing
+ */
+public class DelayedDataDetector {
+
+    private static final String DATE_BUCKETS = "date_buckets";
+    private final long bucketSpan;
+    private final long window;
+    private final DatafeedConfig datafeedConfig;
+    private final Client client;
+    private final Job job;
+
+    public DelayedDataDetector(Job job, DatafeedConfig datafeedConfig, TimeValue window, Client client) {
+        this.job = job;
+        this.bucketSpan = job.getAnalysisConfig().getBucketSpan().millis();
+        this.datafeedConfig = datafeedConfig;
+        long windowMillis = window.millis();
+        if (windowMillis < bucketSpan) {
+            throw new IllegalArgumentException("[window] must be greater or equal to the [bucket_span]");
+        }
+        this.window = windowMillis;
+        this.client = client;
+    }
+
+    public List<BucketWithMissingData> missingData(long latestFinalizedBucket) {
+        final long end = Intervals.alignToFloor(latestFinalizedBucket, bucketSpan);
+        final long start = Intervals.alignToFloor(latestFinalizedBucket - window, bucketSpan);
+        List<Bucket> finalizedBuckets = checkBucketEvents(start, end);
+        Map<Long, Long> indexedData = checkTrueData(start, end);
+        return finalizedBuckets.stream()
+            .filter(bucket -> calculateMissing(indexedData, bucket) > 0)
+            .map(bucket -> BucketWithMissingData.fromMissingAndBucket(calculateMissing(indexedData, bucket), bucket))
+            .collect(Collectors.toList());
+    }
+
+    private List<Bucket> checkBucketEvents(long start, long end) {
+        GetBucketsAction.Request request = new GetBucketsAction.Request(job.getId());
+        request.setStart(Long.toString(start));
+        request.setEnd(Long.toString(end));
+        request.setExcludeInterim(true);
+        request.setPageParams(new PageParams(0, (int)((end - start)/bucketSpan)));
+
+        try (ThreadContext.StoredContext ignore = stashWithOrigin(client.threadPool().getThreadContext(), ML_ORIGIN)) {
+            GetBucketsAction.Response response = client.execute(GetBucketsAction.INSTANCE, request).actionGet();
+            return response.getBuckets().results();
+        }
+    }
+
+    private Map<Long, Long> checkTrueData(long start, long end) {
```

Comment:
Rather than `True` maybe this method could be called `checkCurrentBucketEventCount`?  There's nothing to say it couldn't change again after this method does its work.